File: year2024/day05/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: string[] = file.toString().split('\n');
const [r, u] = [input.slice(0, input.indexOf('')).map(line => line.split('|').map(Number) as [number, number]), input.slice(input.indexOf('') + 1).map(line => line.split(',').map(Number))];
const isUpdateValid = (update: number[], rules: [number, number][]): boolean => !update.some((_, i) => update.slice(i + 1).some((e, j) => rules.some(([a, b]) => a === e && b === update[i])));
const findMiddlePage = (update: number[]): number => update[Math.floor(update.length / 2)];
const reorderUpdate = (update: number[], rules: [number, number][]) => orderRules(rules.filter(([a, b]) => update.includes(a) && update.includes(b)));
const orderRules = (rules: [number, number][]): number[] => {
  const [g, c, n] = [{}, {}, new Set<number>()];
  rules.forEach(([a, b]) => {
    g[a] = (g[a] || []).concat(b);
    c[b] = (c[b] || 0) + 1;
    n.add(a).add(b);
  });
  const [q, r] = [[...n].filter(n => !c[n]), []];
  while (q.length) {
    r.push(q.shift()!);
    (g[r[r.length - 1]] || []).forEach(n => { if (--c[n] === 0) q.push(n) });
  }
  return r;
};
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return u.reduce((out, update) => isUpdateValid(update, r) ? out + findMiddlePage(update) : out, 0);
}

export function part_2(): number {
  return u.reduce((out, update) => isUpdateValid(update, r) ? out : out + findMiddlePage(reorderUpdate(update, r)), 0);
}

File: year2024/day05/python/solution.py
from collections import defaultdict

def main():
  print(part_1())
  print(part_2())


def part_1():
  return sum(int(i[len(i) // 2]) if not any(a in i and b in i and i.index(a) > i.index(b) for a, b in rules) else 0 for i in updates)


def part_2():
  dict = defaultdict(set)
  for a, b in rules:
    dict[a].add(b)
  ans = 0
  for i in updates:
    if any(a in i and b in i and i.index(a) > i.index(b) for a, b in rules):
      i = sorted({j: dict[j] & set(i) for j in i}, key=lambda k: len({j: dict[j] & set(i) for j in i}[k]), reverse=True)
      ans += int(i[len(i) // 2])
  return ans


def setup(path):
  global rules, updates

  with open(path, 'r') as file:
    rules, updates = file.read().split("\n\n")
    rules = [i.split("|") for i in rules.splitlines()]
    updates = [i.split(",") for i in updates.splitlines()]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()